eo-runtime/src/main/java/EOorg/EOeolang/EOcage.java

```diff
@@ -108,4 +112,96 @@ public Phi lambda() {
             return new Data.ToPhi(true);
         }
     }
+
+    /**
+     * Class to trace if the cage got into recursion during the dataization.
+     * @since 0.36
+     * @todo #2836:60min Add a new parameter of recursion depth. This parameter
+     *   should be set by user via pom.xml. We can make DATAIZING_CAGES a
+     *   ConsurrentHashMap which will add thread safety to the class.
+     */
+    private static final class PhTracedEnclosure implements Phi {
+
+        /**
+         * Cages that are currently dataizing. If one cage is datazing and
+         * it needs to be dataized inside current dataizing, the cage will be here,
+         */
+        private static final Set<Integer> DATAIZING_CAGES = new HashSet<>();
```

Comment:
@levBagryansky Since you are using a static variable here, it might cause some problems in a multi-threaded environment. It looks like `ConcurrentHashMap` won't help you here. Perhaps [ThreadLocal](https://docs.oracle.com/javase/8/docs/api/java/lang/ThreadLocal.html) is a more suitable solution here? What do you think?

Reply:
@volodya-lombrozo `ThreadLocal` here can lead to really complex scenarios. What will be with the object when one thread joins to other? Is it OK in our case?  According to documentation
 `ThreadLocal instances are typically private static fields in classes that wish to associate state with a thread` this is not our case.

Reply:
@levBagryansky I'm not insist on `ThreadLocal` here, but `ConcurrentHashMap` seems a strange choice to me. Let's imagine we started dataization in two separate threads, in this case values in `ConcurrentHashMap` might overlap which might cause problems.  I suggest to discuss the problem in a separate `todo`.